File: AngularJSDemo/AngularJSDemo.cs
using Bridge;
using Bridge.AngularJS;
using Bridge.Html5;
using System;
using System.Collections.Generic;

namespace AngularJSDemo
{
    /// <summary>
    /// Checkpoints:
    /// Alpha Bravo Charlie Delta Echo Foxtrot Golf Hotel India
    /// Juliet Kilo Lima Mike November Oscar Papa Quebec Romeo
    /// Sierra Tango Uniform Victor Whiskey X-ray Yankee Zulu
    /// http://home.earthlink.net/~malcolmhamer/alpha.htm
    /// </summary>
    public class AngularJSDemo
    {

        public class ControllerDataObjectStructure
        {
            public string Message;
            public string Foo;
            public string Bar;
            public object[] Checkpoints;
        }

        static AngularJSDemo()
        {
            Console.Log("Checkpoint Yankee: [ctor] nothng at this point will " +
                "like AngularJS.");
        }

        public static Module hwbApp;

        /// <summary>
        /// AngularJS requires to be defined on document parse.
        /// TODO: create an onParse attribute
        /// </summary>
        public static void StartUp()
        {
            Console.Log("Checkpoint Charlie: I must be printed BEFORE " +
                "Checkpoint 1.");

            AngularJSDemo.hwbApp = Angular.Module("hwbApp");

            var controllerData = new { message = "Hello, AngularJS message " +
                    "defined in Bridge's C#! :D" };
            AngularJSDemo.hwbApp.Controller("hwbctl", controllerData);
            AngularJSDemo.hwbApp.Directive("brdEntryPoint",
                AngularJSDemo.dynMehTemplate);

            var checkpoints = new List<object> {
                new { callsign = "Alpha", id = 98 },
                new { callsign = "Baker", id = 78 },
                new { callsign = "Charlie", id = 58 },
                new { callsign = "Delta", id = 9 }
            };

            var controllerStrongData = new ControllerDataObjectStructure()
            {
                Message = "Hello, AngularJS message defined in Bridge's C#'s " +
                    "strongly typed class! :D",
                Foo = "Foo fighters from strong C#.",
                Bar = "Strong C# 777 slot",
                Checkpoints = checkpoints.ToArray()
            };

            AngularJSDemo.hwbApp.Controller("hwbSctl", controllerStrongData);
            AngularJSDemo.hwbApp.Directive("brdEntryPointForThree",
                AngularJSDemo.ThreeWayFunction);

            Document.DocumentElement.SetNGApp(AngularJSDemo.hwbApp);
        }

        /// <summary>
        /// This function determines the format of the contents of the entry
        /// point where the dynamic code will be injected.
        /// </summary>
        /// <returns></returns>
        public static object dynMehTemplate()
        {
            var span = new SpanElement();
            span.SetNGController("hwbctl");
            span.InnerHTML = "AJS says by directive: [{{message}}]";
            return new { template = span.OuterHTML };
        }
        public static object ThreeWayFunction()
        {
            var span = new SpanElement();
            span.SetNGController("hwbSctl");
            span.InnerHTML = "AJS for three scopined: " +
                "[msg:{{message}}][foo:{{foo}}][bar:{{bar}}]";
            return new { template = span.OuterHTML };
        }

        // This one must be called after the tag has been defined in the
        // HTML DOM, yet before HTML DOM has been loaded (DOMContentLoaded)
        public static void StartUpIdPhase()
        {
            if (AngularJSDemo.hwbApp == null) {
                AngularJSDemo.StartUp();
            }

            var x = Document.GetElementById("entryPoint");
            if (x != null)
            {
                //x.SetAttribute("ng-controller", "hwbctl"); // same as below
                x.SetNGController("hwbctl");
                x.InnerHTML = "AJS says by element ID: [{{message}}]";
            }

            Document.Body.AppendChild(AngularJSDemo.GetRepeatRegion());

            // Broken :(
            AngularJSDemo.DefineNiceController();
        }

        public static Element GetRepeatRegion() {
            var itemsSpan = new SpanElement();
            itemsSpan.InnerHTML = "Checkpoint";
            var itemsPara = new ParagraphElement();
            itemsPara.InnerHTML = "{{checkpoint.callsign}}[{{checkpoint.id}}]";

            var itemsLI = new LIElement();
            itemsLI.SetNGRepeat("checkpoint", "checkpoints");
            itemsLI.AppendChild(itemsSpan);
            itemsLI.AppendChild(itemsPara);

            var itemsUL = new UListElement();
            itemsUL.AppendChild(itemsLI);

            var itemsSubSpan = new SpanElement()
            {
                InnerHTML = "[{{checkpoint.callsign}}.{{checkpoint.id}}] "
            };

            var itemsSearchBox = new InputElement();
            itemsSearchBox.SetNGModel("cpFilter");

            var itemsOrderSelector = GetOrderBySelector("cpOrderBy");

            itemsSubSpan.SetNGRepeat("checkpoint", "checkpoints",
                itemsSearchBox.GetNGModel(), itemsOrderSelector.GetNGModel());

            var itemsDiv = new DivElement();
            itemsDiv.SetNGController("hwbSctl");
            itemsDiv.AppendChild(itemsUL);
            itemsDiv.AppendChild(itemsSearchBox);
            itemsDiv.AppendChild(itemsOrderSelector);
            itemsDiv.AppendChild(itemsSubSpan);

            return itemsDiv;
        }

        public static Element GetOrderBySelector(string ngModelName)
        {
            var itemsOrderSelector = new SelectElement();
            itemsOrderSelector.SetNGModel(ngModelName);

            itemsOrderSelector.Add(new OptionElement() {
                Value = "",
                InnerHTML = "No special order"
            });

            itemsOrderSelector.Add(new OptionElement() {
                Value = "name",
                InnerHTML = "CallSign"
            });

            itemsOrderSelector.Add(new OptionElement() {
                Value = "id",
                InnerHTML = "Internal"
            });

            return itemsOrderSelector;
        }

        public static void DefineNiceController()
        {
            AngularJSDemo.hwbApp.Controller<ControllerDataObjectStructure>
                ("hwcSctl", CtlFunction);

            var ctlDiv = new DivElement();
            ctlDiv.SetNGController("hwcSctl");
            Document.Body.AppendChild(ctlDiv);

            var fltFld = new InputElement();
            fltFld.SetNGModel("hwcFlt");
            ctlDiv.AppendChild(fltFld);

            var ordFld = new SelectElement();
            ordFld.SetNGModel("hwcOrderBy");
            ordFld.Add(new OptionElement() {
                Value = "Checkpoint",
                InnerHTML = "Alphabetically"
            });
            ordFld.Add(new OptionElement() {
                Value = "id",
                InnerHTML = "Series ID"
            });
            ctlDiv.AppendChild(ordFld);

            var rptSpan = new SpanElement();
            rptSpan.SetNGRepeat("checkpoint", "checkpoints", fltFld.GetNGModel(),
                ordFld.GetNGModel());
            rptSpan.InnerHTML = "{{checkpoint.callsign}}[{{checkpoint.id}}] ";
            ctlDiv.AppendChild(rptSpan);
        }

        public static void CtlFunction(
            [Name("$scope")] ControllerDataObjectStructure scope)
        {
            scope.Message = "Hello, Angular defined as nice $scope function.";
            scope.Foo = "Foo fighter nice $scope";
            scope.Bar = "Nice $scope 777 slot";
            scope.Checkpoints = new List<object> {
                new { callsign = "Alpha", id = 85 },
                new { callsign = "Baker", id = 35 },
                new { callsign = "Charlie", id = 55 },
                new { callsign = "Delta", id = 6 }
            }.ToArray();
        }

        /// <summary>
        /// Trying to define AngularJS controller/app after DOM is loaded will
        /// not work. Angular is supposed to be defined during the page
        /// composition, not after it is ready.
        /// </summary>
        [Ready]
        public static void UpdateControls()
        {
            Console.Log("Checkpoint Zebra: [rdy] nothing at this point will " +
                "like AngularJS.");

            var lbl = new SpanElement() { InnerHTML = "We have Bridge script " +
                    "running." };
            Document.Body.AppendChild(lbl);
        }
    }
}

File: AngularJSDemo/js/angularJSDemo.js
/* global Bridge */

"use strict";

/** @namespace AngularJSDemo */

/**
 * Checkpoints:
 Alpha Bravo Charlie Delta Echo Foxtrot Golf Hotel India
 Juliet Kilo Lima Mike November Oscar Papa Quebec Romeo
 Sierra Tango Uniform Victor Whiskey X-ray Yankee Zulu
 http://home.earthlink.net/~malcolmhamer/alpha.htm
 *
 * @public
 * @class AngularJSDemo.AngularJSDemo
 */
Bridge.define('AngularJSDemo.AngularJSDemo', {
    statics: {
        constructor: function () {
            console.log("Checkpoint Yankee: [ctor] nothng at this point will like AngularJS.");
        },
        hwbApp: null,
        config: {
            init: function () {
                Bridge.ready(this.updateControls);
            }
        },
        /**
         * AngularJS requires to be defined on document parse.
         TODO: create an onParse attribute
         *
         * @static
         * @public
         * @this AngularJSDemo.AngularJSDemo
         * @memberof AngularJSDemo.AngularJSDemo
         * @return  {void}        
         */
        startUp: function () {
            console.log("Checkpoint Charlie: I must be printed BEFORE Checkpoint 1.");

            AngularJSDemo.AngularJSDemo.hwbApp = angular.module("hwbApp", []);

            var controllerData = { message: "Hello, AngularJS message defined in Bridge's C#! :D" };
            AngularJSDemo.AngularJSDemo.hwbApp.controller("hwbctl", ['$scope', function ($scope) { for (var key in controllerData) $scope[key] = controllerData[key] }]);
            AngularJSDemo.AngularJSDemo.hwbApp.directive("brdEntryPoint", AngularJSDemo.AngularJSDemo.dynMehTemplate);

            var checkpoints = Bridge.merge(new Bridge.List$1(Object)(), [
                [{ callsign: "Alpha", id: 98 }], 
                [{ callsign: "Baker", id: 78 }], 
                [{ callsign: "Charlie", id: 58 }], 
                [{ callsign: "Delta", id: 9 }]
            ] );

            var controllerStrongData = Bridge.merge(new AngularJSDemo.AngularJSDemo.ControllerDataObjectStructure(), {
                message: "Hello, AngularJS message defined in Bridge's C#'s strongly typed class! :D", 
                foo: "Foo fighters from strong C#.", 
                bar: "Strong C# 777 slot", 
                checkpoints: checkpoints.toArray()
            } );

            AngularJSDemo.AngularJSDemo.hwbApp.controller("hwbSctl", ['$scope', function ($scope) { for (var key in controllerStrongData) $scope[key] = controllerStrongData[key] }]);
            AngularJSDemo.AngularJSDemo.hwbApp.directive("brdEntryPointForThree", AngularJSDemo.AngularJSDemo.threeWayFunction);

            document.documentElement.setAttribute('ng-app', AngularJSDemo.AngularJSDemo.hwbApp.name);
        },
        /**
         * This function determines the format of the contents of the entry
         point where the dynamic code will be injected.
         *
         * @static
         * @public
         * @this AngularJSDemo.AngularJSDemo
         * @memberof AngularJSDemo.AngularJSDemo
         * @return  {Object}        
         */
        dynMehTemplate: function () {
            var span = document.createElement('span');
            span.setAttribute('ng-controller', "hwbctl");
            span.innerHTML = "AJS says by directive: [{{message}}]";
            return { template: span.outerHTML };
        },
        threeWayFunction: function () {
            var span = document.createElement('span');
            span.setAttribute('ng-controller', "hwbSctl");
            span.innerHTML = "AJS for three scopined: [msg:{{message}}][foo:{{foo}}][bar:{{bar}}]";
            return { template: span.outerHTML };
        },
        startUpIdPhase: function () {
            if (AngularJSDemo.AngularJSDemo.hwbApp === null) {
                AngularJSDemo.AngularJSDemo.startUp();
            }

            var x = document.getElementById("entryPoint");
            if (x !== null) {
                //x.SetAttribute("ng-controller", "hwbctl"); // same as below
                x.setAttribute('ng-controller', "hwbctl");
                x.innerHTML = "AJS says by element ID: [{{message}}]";
            }

            document.body.appendChild(AngularJSDemo.AngularJSDemo.getRepeatRegion());

            // Broken :(
            AngularJSDemo.AngularJSDemo.defineNiceController();
        },
        getRepeatRegion: function () {
            var itemsSpan = document.createElement('span');
            itemsSpan.innerHTML = "Checkpoint";
            var itemsPara = document.createElement('p');
            itemsPara.innerHTML = "{{checkpoint.callsign}}[{{checkpoint.id}}]";

            var itemsLI = document.createElement('li');
            itemsLI.setAttribute('ng-repeat', '' + "checkpoint" + ' in ' + "checkpoints");
            itemsLI.appendChild(itemsSpan);
            itemsLI.appendChild(itemsPara);

            var itemsUL = document.createElement('ul');
            itemsUL.appendChild(itemsLI);

            var itemsSubSpan = Bridge.merge(document.createElement('span'), {
                innerHTML: "[{{checkpoint.callsign}}.{{checkpoint.id}}] "
            } );

            var itemsSearchBox = document.createElement('input');
            itemsSearchBox.setAttribute('ng-model', "cpFilter");

            var itemsOrderSelector = AngularJSDemo.AngularJSDemo.getOrderBySelector("cpOrderBy");

            itemsSubSpan.setAttribute('ng-repeat', '' + "checkpoint" + ' in ' + "checkpoints" + ' | filter: ' + itemsSearchBox.getAttribute('ng-model') + ' | orderBy: ' + itemsOrderSelector.getAttribute('ng-model'));

            var itemsDiv = document.createElement('div');
            itemsDiv.setAttribute('ng-controller', "hwbSctl");
            itemsDiv.appendChild(itemsUL);
            itemsDiv.appendChild(itemsSearchBox);
            itemsDiv.appendChild(itemsOrderSelector);
            itemsDiv.appendChild(itemsSubSpan);

            return itemsDiv;
        },
        getOrderBySelector: function (ngModelName) {
            var itemsOrderSelector = document.createElement('select');
            itemsOrderSelector.setAttribute('ng-model', ngModelName);

            itemsOrderSelector.add(Bridge.merge(document.createElement('option'), {
                value: "", 
                innerHTML: "No special order"
            } ));

            itemsOrderSelector.add(Bridge.merge(document.createElement('option'), {
                value: "name", 
                innerHTML: "CallSign"
            } ));

            itemsOrderSelector.add(Bridge.merge(document.createElement('option'), {
                value: "id", 
                innerHTML: "Internal"
            } ));

            return itemsOrderSelector;
        },
        defineNiceController: function () {
            AngularJSDemo.AngularJSDemo.hwbApp.controller("hwcSctl", AngularJSDemo.AngularJSDemo.ctlFunction);

            var ctlDiv = document.createElement('div');
            ctlDiv.setAttribute('ng-controller', "hwcSctl");
            document.body.appendChild(ctlDiv);

            var fltFld = document.createElement('input');
            fltFld.setAttribute('ng-model', "hwcFlt");
            ctlDiv.appendChild(fltFld);

            var ordFld = document.createElement('select');
            ordFld.setAttribute('ng-model', "hwcOrderBy");
            ordFld.add(Bridge.merge(document.createElement('option'), {
                value: "Checkpoint", 
                innerHTML: "Alphabetically"
            } ));
            ordFld.add(Bridge.merge(document.createElement('option'), {
                value: "id", 
                innerHTML: "Series ID"
            } ));
            ctlDiv.appendChild(ordFld);

            var rptSpan = document.createElement('span');
            rptSpan.setAttribute('ng-repeat', '' + "checkpoint" + ' in ' + "checkpoints" + ' | filter: ' + fltFld.getAttribute('ng-model') + ' | orderBy: ' + ordFld.getAttribute('ng-model'));
            rptSpan.innerHTML = "{{checkpoint.callsign}}[{{checkpoint.id}}] ";
            ctlDiv.appendChild(rptSpan);
        },
        ctlFunction: function ($scope) {
            $scope.message = "Hello, Angular defined as nice $scope function.";
            $scope.foo = "Foo fighter nice $scope";
            $scope.bar = "Nice $scope 777 slot";
            $scope.checkpoints = Bridge.merge(new Bridge.List$1(Object)(), [
                [{ callsign: "Alpha", id: 85 }], 
                [{ callsign: "Baker", id: 35 }], 
                [{ callsign: "Charlie", id: 55 }], 
                [{ callsign: "Delta", id: 6 }]
            ] ).toArray();
        },
        /**
         * Trying to define AngularJS controller/app after DOM is loaded will
         not work. Angular is supposed to be defined during the page
         composition, not after it is ready.
         *
         * @static
         * @public
         * @this AngularJSDemo.AngularJSDemo
         * @memberof AngularJSDemo.AngularJSDemo
         * @return  {void}        
         */
        updateControls: function () {
            console.log("Checkpoint Zebra: [rdy] nothing at this point will like AngularJS.");

            var lbl = Bridge.merge(document.createElement('span'), {
                innerHTML: "We have Bridge script running."
            } );
            document.body.appendChild(lbl);
        }
    }
});

Bridge.define('AngularJSDemo.AngularJSDemo.ControllerDataObjectStructure', {
    message: null,
    foo: null,
    bar: null,
    checkpoints: null
});

